va/va_prot.h

```diff
@@ -0,0 +1,353 @@
+ *     assert(0);
+ * }
+ * \endcode
+ *
+ * \section api_prot_setup Set up a protected session
+ *
+ * A protected session is a session between application and GPU hardware.
+ * This protected session can be attached to the display. Also with protected session,
+ * application can do specific TEE execution functionality with TEE inside GPU, but it depends
+ * on IHV's implementation.
+ * Before creating a protected session, it needs to create a config first via
+ * vaCreateConfig(). Then using this config id to create a protected session via
+ * vaCreateProtectedSession().
+ *
+ * The general control flow is demonstrated by the following pseudo-code:
+ * \code
+ * // Create config
+ * VAConfigID config_id;
+ *
+ * wired.bits.session_type = VA_PROTECTED_DISPLAY_LINK_SESSION_TYPE0;
+ * wired.bits.protocol_version = VA_PROTECTED_DISPLAY_LINK_HDCP_VERSION_22;
+ * attrib_prot[0].value = wired.value;
+ * va_status = vaCreateConfig(va_dpy, VAProfileProtected, VAEntrypointProtectedDisplayLink, attrib_prot, 1, &config_id);
+ * CHECK_VASTATUS(va_status, "vaCreateConfig");
+ * \endcode
+ *
+ * Once the config is set up, we can create protected session via vaCreateProtectedSession().
+ * \code
+ * // Create a protected session
+ * VAProtectedSessionID protected_session;
+ *
+ * va_status = vaCreateProtectedSession(
+ *     va_dpy,
+ *     config_id,
+ *     &protected_session);
+ * CHECK_VASTATUS(va_status, "vaCreateProtectedSession");
+ * \endcode
+ *
+ * \section api_prot_attach Attach/Detach protected session to the display which want to enable/disable protection
+ *
+ * Protected session is attached to display to enable protection.
+ *
+ * \code
+ * vaAttachProtectedSession(va_dpy, 0, protected_session);
+ * foreach (iteration) {
+ *     vaBeginPicture(va_dpy, decode_ctx, surface);
+ *     ...
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id1, 1);
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id2, 1);
+ *     ...
+ *     vaEndPicture(va_dpy, decode_ctx);
+ * }
+ * vaDetachProtectedSession(va_dpy, 0);
+ * \endcode
+ *
+ * \section api_prot_update Updating protected session hardware status
+ *
+ * A function for updating hardware status of the protected session. Sometimes the status is
+ * unknown at protected session creation time. When the information is gathered, we can update
+ * the hardware status via vaProtectedSessionHwUpdate().
+ *
+ * \code
+ * INFO *info = &oem_info_gathered;
+ * VABufferID buffer;
+ * VAProtectedSessionHwUpdateBuffer update_buff = {0};
+ *
+ * update_buff.input.input_data = info;
+ * update_buff.input.data_size = sizeof(info);
+ * va_status = vaCreateBuffer(
+ *                 va_dpy,
+ *                 protected_session,
+ *                 (VABufferType) VAProtectedSessionHwUpdateBufferType,
+ *                 sizeof(update_buff),
+ *                 1,
+ *                 &update_buff,
+ *                 &buffer);
+ *
+ * va_status = vaProtectedSessionHwUpdate(va_dpy, protected_session, buffer);
```

Comment:
You gave an example above on how to use protected sessions:
```
 * vaAttachProtectedSession(va_dpy, 0, protected_session);
 * foreach (iteration) {
 *     vaBeginPicture(va_dpy, decode_ctx, surface);
 *     ...
 *     vaRenderPicture(va_dpy, decode_ctx, &buf_id1, 1);
 *     vaRenderPicture(va_dpy, decode_ctx, &buf_id2, 1);
 *     ...
 *     vaEndPicture(va_dpy, decode_ctx);
 * }
 * vaDetachProtectedSession(va_dpy, 0);
```

But reading your comments about vaProtectedSessionHwUpdate and vaProtectedSessionExecute I completely don't understand why they are actually needed! From description you have - remove these 2 things and everything will still work in your example above.

Reply:
Thanks for your feedback. I restructure and refine the comments and updated. Please check if it is better or not.

Reply:
With your last change it seems you indicate that hwupdate should be called before attaching protected session. Is that right? I wonder are there any usage models which need exec() and hwupdate() calls after the attach?

Reply:
HwUpdate related definitions are removed.